### PR TITLE
Implementing Andy Gio's mssql fix

### PR DESCRIPTION
**Pull request by roya0045**
Implementing the suggested fix by Andy Gio in https://issues.qgis.org/issues/20592

## Description
This fix should allow the use of triggers on INSERT in MSSQL while maintaining the current behaviour.

To do so a temporary table is created in order to return the ID of the modified features instead of directly using OUTPUT to do so.

This fix has not been tested in my fork.
The prepare commit script has not been run since this is a minor change.

All credit should go to Mr. Andy Gio for this code that was provided in : https://issues.qgis.org/issues/20592.

### CODE REVIEW

File: src/providers/mssql/qgsmssqlprovider.cpp [hunk 1]
```diff
@@ -947,10 +952,14 @@ bool QgsMssqlProvider::addFeatures( QgsFeatureList &flist, Flags flags )
     statement += QStringLiteral( ") " );
     if ( !( flags & QgsFeatureSink::FastInsert ) )
     {
-      statement += QStringLiteral( " OUTPUT inserted." ) + mFidColName;
+      statement += QStringLiteral( " OUTPUT inserted." ) + mFidColName + QStringLiteral( "  INTO @px " );
```
Comment (m-kuhn) [inline src/providers/mssql/qgsmssqlprovider.cpp]: I don't know mysql very well. Should inserted and the column name be quoted (something like "inserted"."colname")?

File: src/providers/mssql/qgsmssqlprovider.cpp [hunk 2]
```diff
@@ -947,10 +952,14 @@ bool QgsMssqlProvider::addFeatures( QgsFeatureList &flist, Flags flags )
...
     }
     statement += QStringLiteral( " VALUES (" ) + values + ')';
-
+    
+    if ( !( flags & QgsFeatureSink::FastInsert ) )
+    {    
+        statement += '; SELECT id FROM @px;';
```
Comment (m-kuhn) [inline src/providers/mssql/qgsmssqlprovider.cpp]:
```suggestion
         statement += QStringLiteral( "; SELECT id FROM @px;" );
```

**Comment by m-kuhn** [comment]
Thank you for opening this pull request.
Did you compile and verify this fix? It doesn't compile in the current shape (see my suggestion) and I cannot verify the implementation because I don't have a mssql environment running here.

**Comment by roya0045** [comment]
I did not compile it. But I did some testing on a mssql environment prior to implementing this solution and the behaviour was the same as the current one without causing any issues when there are triggers on INSERT.

Thank you for the fix this was an oversight on my part.

File: src/providers/mssql/qgsmssqlprovider.cpp (hunk 1 above)
Comment (roya0045) [reply src/providers/mssql/qgsmssqlprovider.cpp]: Quotations are not necessary but MSSQL likes when databases schema and tables are between bracket but the inserted table can be referenced without quotation and other such.

Here is an example with OUTPUT https://docs.microsoft.com/en-us/sql/t-sql/queries/output-clause-transact-sql?view=sql-server-2017#examples

Comment (m-kuhn) [reply src/providers/mssql/qgsmssqlprovider.cpp]: Actually I think there should be a `QgsMssqlProvider::quotedIdentifier` method doing the same thing as the [`QUOTENAME`](https://docs.microsoft.com/en-us/sql/t-sql/functions/quotename-transact-sql?view=sql-server-2017) function. But that's out of scope for this pull request.

**Comment by m-kuhn** [comment]
Looking good to me now.
Would be nice to get a pair of eyes from @nyalldawson @NathanW2 too.

**Comment by m-kuhn** [comment]
No reason to merge master back into this pull request as long as there are no merge conflicts ;)

**Comment by nyalldawson** [comment]
I'll test this next week when I'm back in the office and have access to a SQL Server instance

**Comment by roya0045** [comment]
@m-kuhn Nah I'm just not too used to github and wanted to keep my branch up to date for the other project and it just added this to this pull since it was my master, I should have branched it in restrospective.

**Comment by m-kuhn** [comment]
I see, it's also nothing wrong with it, just wanted to let you know that it's unrequired, all good :)

**Comment by roya0045** [comment]
I'll reopen it it a new request I'm breaking my own repo too  much.

**Pull request closed without merging**